ci: Use macos-14 GHA image (x86_64-apple-darwin22.6.0 -> arm64-apple-darwin23.6.0)

maflcko opened the pull request:
There shouldn't be any downside, because XCode remains pinned to the same version.

However, builds are expected to be a bit faster with M1, which seems nice.

Commit:
ci: Use macos-14 GHA image

willcl-ark commented on .github/workflows/ci.yml:
```diff
@@ -74,11 +74,11 @@ jobs:
 
-  macos-native-x86_64:
-    name: 'macOS 13 native, x86_64, no depends, sqlite only, gui'
+  macos-native-arm64:
```
This is still building using x86_64-apple-darwin as the host (from 00_setup_env_mac_native.sh). I wonder if the "native" here is therefore confusing?

willcl-ark reviewed (commented):
Concept ACK.

Very much in favour of speeding these jobs up! Changes look good to me so far, just one Q on the jobname with this change in architecture.

Commit:
ci: Remove incorrectly hardcoded HOST in mac_native task

The HOST is wrong, because this is a native build, where the HOST is
supposed to be auto-detected.

hebasto reviewed (commented):
Concept ACK.

Perhaps explicitly mention the switch from the `x86_64` architecture to `arm64`?

hebasto reviewed (approved):
ACK fab932b4211ea0a4c10e484c08042552f9c58c15.

We were already using an `arm64` CI task before https://github.com/bitcoin/bitcoin/pull/28187.

willcl-ark reviewed (approved):
ACK fab932b4211ea0a4c10e484c08042552f9c58c15
